compiler/rustc_codegen_ssa/src/mir/naked_asm.rs

```diff
@@ -104,6 +106,215 @@ fn inline_to_global_operand<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
+            if sess.target.pointer_width == 32 {
+                Architecture::Aarch64_Ilp32
+            } else {
+                Architecture::Aarch64
+            },
+            None,
+        ),
+        "x86" => (Architecture::I386, None),
+        "s390x" => (Architecture::S390x, None),
+        "mips" | "mips32r6" => (Architecture::Mips, None),
+        "mips64" | "mips64r6" => (Architecture::Mips64, None),
+        "x86_64" => (
+            if sess.target.pointer_width == 32 {
+                Architecture::X86_64_X32
+            } else {
+                Architecture::X86_64
+            },
+            None,
+        ),
+        "powerpc" => (Architecture::PowerPc, None),
+        "powerpc64" => (Architecture::PowerPc64, None),
+        "riscv32" => (Architecture::Riscv32, None),
+        "riscv64" => (Architecture::Riscv64, None),
+        "sparc" => {
+            if sess.unstable_target_features.contains(&sym::v8plus) {
+                // Target uses V8+, aka EM_SPARC32PLUS, aka 64-bit V9 but in 32-bit mode
+                (Architecture::Sparc32Plus, None)
+            } else {
+                // Target uses V7 or V8, aka EM_SPARC
+                (Architecture::Sparc, None)
+            }
+        }
+        "sparc64" => (Architecture::Sparc64, None),
+        "avr" => (Architecture::Avr, None),
+        "msp430" => (Architecture::Msp430, None),
+        "hexagon" => (Architecture::Hexagon, None),
+        "bpf" => (Architecture::Bpf, None),
+        "loongarch64" => (Architecture::LoongArch64, None),
+        "csky" => (Architecture::Csky, None),
+        "arm64ec" => (Architecture::Aarch64, Some(SubArchitecture::Arm64EC)),
+
+        // added here
+        "wasm32" => (Architecture::Wasm32, None),
+        "wasm64" => (Architecture::Wasm64, None),
+        "m68k" => (Architecture::M68k, None),
+
+        // Unsupported architecture.
+        _ => return None,
+    };
+
+    Some((architecture, subarchitecture))
+}
+
+/// Enable the function's target features in the body of the function, then disable them again
+fn enable_disable_target_features<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    attrs: &CodegenFnAttrs,
+) -> Option<(String, String)> {
+    use std::fmt::Write;
+
+    let mut begin = String::new();
+    let mut end = String::new();
+
+    let (architecture, _subarchitecture) = parse_architecture(tcx.sess)?;
+    let features = attrs.target_features.iter().filter(|attr| !attr.implied);
+
+    match architecture {
+        Architecture::X86_64 | Architecture::X86_64_X32 => { /* do nothing */ }
+
+        Architecture::Aarch64 | Architecture::Aarch64_Ilp32 | Architecture::Arm => {
+            // https://developer.arm.com/documentation/100067/0611/armclang-Integrated-Assembler/AArch32-Target-selection-directives?lang=en
+
+            for feature in features {
+                writeln!(begin, ".arch_extension {}", feature.name).unwrap();
+
+                writeln!(end, ".arch_extension no{}", feature.name).unwrap();
```

Comment:
[AFAIK, arch_extension is stateless](https://github.com/taiki-e/portable-atomic/blob/75a36c33b38c4c68f4095e95f106cfbedce9a914/src/imp/atomic128/aarch64.rs#L330), so no* is probably unneeded.

Reply:
for LLVM that's fine but for other backends it might not be. I've left it in just to be defensive.

Reply:
FYI, I also checked GCC, as mentioned in the linked comment.

> ```rs
> // (It seems GCC effectively ignores this directive and always allow FEAT_LSE instructions: https://godbolt.org/z/W9W6rensG)
> ```

Reply:
right, I was thinking of cranelift now, and maybe the WIP C and C# backends (if that does asm) in the future

Reply:
As for cranelift backend, it [uses LLVM by default and GNU AS by option](https://github.com/rust-lang/rustc_codegen_cranelift/blob/30ece8da06882b9cbca9f7e73f6804c669449a17/src/global_asm.rs#L154), so there should be no problem.

Also, if there is an assembler that treats arch_extension as stateful, I think ending with `.arch_extension no*` is also wrong, as mentioned in riscv's case.

Reply:
yes, there doesn't seem to be a similar push/pop mechanism either. I've removed that line and left a comment linking to your comment.

Reply:
Sorry, in `asm!` that I tested in my project, `.arch_extension` is stateless, but in `global_asm!` that used in `naked_asm!`, it seems to be stateful.
https://godbolt.org/z/4Y6xTsdYE

(However, as mentioned above, ending with `.arch_extension no*` is also wrong if that function is already enabled...)

Reply:
ah, that is surprising. Maybe we can look at the global target features and then now enable/disable any features that are already globally enabled? I'll give that a go

Reply:
the surprising part to me is the difference in behavior between `asm!` and `global_asm!`, btw

Reply:
> Maybe we can look at the global target features and then now enable/disable any features that are already globally enabled?

That sounds like a reasonable approach!

Reply:
I believe this works. I'm testing both that global features are not disabled, and that target features don't leak out of a naked function.